_posts/2022-11-11-functional-crystal.md

```diff
@@ -0,0 +1,428 @@
+
+```cr
+abstract class Ast
+  def inspect(io)
+    case self
+    in BinOp
+      self.left.inspect io
+      self.operator.to_s io
+      self.right.inspect io
+    in IntLiteral
+      self.number.to_s io
+    end
+  end
+end
+```
+
+Note how this method is more concise than the OOP one (there's a little white lie, see the coming section). Also importantly, note how we're casing on the type of `self`.
+
+## The missing bits
+
+We've seen that Crystal has good support for several functional programming patterns. But there are a few improvements that can be made in order to improve it.
+
+### A more functional stdlib
+
+The stdlib pervasively raises exceptions, and besides nilable types, has little support for exceptions-as-values.
+
+### Better type inference for closures
+
+It's possible to create a closure and assign it to a variable with the following syntax:
+
+```cr
+double = ->(x : Int32) { 2 * x }
+
+p 3.try(&double).try(&double) # => 12
+```
+
+Unfortunately, we need to be explicit about the type of the input parameters in the closure (in this case, `Int32`). Crystal won't infer it even if it could be guessed from the context, as with a regular function or method. It's not a critical aspect, but following the same ergonomics of the language also for closures would be a nice touch.
+
+### Generic aliases
+
+Right now, type aliases can't be generic. Therefore, it's currently impossible to define something of the sort:
+
+```cr
+alias Result(T) = T | Exception
+```
+
+### Immutable datatypes
+
+A key aspect of Abstract Data Types we didn't mentioned so far is that they are immutable. But we can't use structs to get immutability, because structs do not allow recursive definitions. That is, the following is invalid:
+
+```cr
+abstract struct Ast
+end
+
+struct BinOp < Ast
+  getter operator : String
+  getter left : Ast, right : Ast
+
+  def initialize(@operator, @left, @right)
+  end
+end
+```
+
+The error is descriptive:
+
+```
+The struct BinOp has, either directly or indirectly,
+an instance variable whose type is, eventually, this same
+struct. This makes it impossible to represent the struct
+in memory, because the size of this instance variable depends
+on the size of this struct, which depends on the size of
+this instance variable, causing an infinite cycle.
+```
+
+A missing piece is that structs are placed in the stack, unlike classes, and that is why the compiler needs to know ahead of time the exact size of it.
+
+If we want to have immutable classes, that are placed in the heap instead of in the stack (and therefore, can have recursive instances), the compiler should add support for it. One way to achieve that is to have the compiler call a method when passing `Reference` objects around, and have `Reference` return `self` and a new `ImmutableReference` type of object return a copy. If it sounds costly, fear not as the LLVM backend should optimize it in `--release` mode.
```

Comment:
```suggestion
If we want to have immutable classes, that are placed in the heap instead of in the stack (and therefore, can represent recursive structures), the compiler should add support for it. One way to achieve that is to have the compiler call a method when passing `Reference` objects around, and have `Reference` return `self` and a new `ImmutableReference` type of object return a copy. If it sounds costly, fear not as the LLVM backend should optimize it in `--release` mode.
```

I don't get this idea about a compiler feature. If you want an immutable class type, you can get that by just not defining any setter or mutating its instance variables in any other way.
This requires discipline when defining types, which might be sub-par. But it's definitely possible to get immutable types even without any additional compiler features.

Reply:
Not sure I follow. Of course you can build stuff for each particular case, but the point of a compiler should be to make it easier for you to code certain patterns. This said, I now disagree with my proposed solution and more so to the fact that I even added a solution. That's not the point of this article 🤦